rust/src/main.rs

```diff
@@ -33,108 +33,124 @@ pub struct EvTag {
-                     try!(self.checksum_submodule(hash, &submodule));
-                 }
-                 ObjectType::Tag => {}
-                 ObjectType::Any => panic!("Found an Any type?")
-             }
-         }
-         Ok(())
-     }
-
-     fn checksum_commit_contents(&self, repo: &Repository, commit: Commit, hash: &mut Hasher) -> Result<(), Error> {
-         try!(self.checksum_object(hash, commit.as_object()));
-         let tree = try!(commit.tree());
-         try!(self.checksum_tree(repo, tree, hash));
-         Ok(())
-     }
-
-     pub fn compute(&self, specified_oid: Oid) -> Result<Vec<u8>, Error> {
-         let mut hash = Hasher::new(openssl::crypto::hash::Type::SHA512);
-         let commit = try!(self.repo.find_commit(specified_oid));
-         try!(self.checksum_commit_contents(&self.repo, commit, &mut hash));
-         Ok(hash.finish())
-     }
+    pub fn new(path: &str) -> Result<EvTag, Error> {
+        Ok(EvTag {
+            repo: Repository::discover(path)?,
+        })
+    }
+
+    fn checksum_object(&self, hash: &mut Hasher, object: &Object) -> Result<(), Error> {
+        let contentbuf = object.object_data()?;
+        let header = format!("{} {}", object.kind().unwrap().str(), contentbuf.len());
+        hash.write(header.as_bytes()).unwrap();
+        let nulbyte: [u8; 1] = [0; 1];
+        hash.write(&nulbyte).unwrap();
+        hash.write(contentbuf).unwrap();
+        Ok(())
+    }
+
+    fn checksum_submodule(&self, hash: &mut Hasher, submodule: &Submodule) -> Result<(), Error> {
+        let subrepo = submodule.open()?;
+        let sub_head = submodule.workdir_id().expect("Failed to find workdir id");
+        let commit = subrepo.find_commit(sub_head)?;
+        self.checksum_commit_contents(&subrepo, commit, hash)?;
+        Ok(())
+    }
+
+    fn checksum_tree(&self, repo: &Repository, tree: Tree, hash: &mut Hasher) -> Result<(), Error> {
+        self.checksum_object(hash, tree.as_object())?;
+        for entry in tree.iter() {
+            match entry
+                .kind()
+                .expect("I have no idea what a None entry would be...")
+            {
+                ObjectType::Blob => {
+                    let object = repo.find_object(entry.id(), entry.kind())?;
+                    self.checksum_object(hash, &object)?
+                }
+                ObjectType::Tree => {
+                    let tree = repo.find_tree(entry.id())?;
+                    self.checksum_tree(repo, tree, hash)?;
+                }
+                ObjectType::Commit => {
+                    let submodule = repo.find_submodule(entry.name().unwrap())?;
+                    self.checksum_submodule(hash, &submodule)?;
+                }
+                ObjectType::Tag => {}
+                ObjectType::Any => panic!("Found an Any type?"),
+            }
+        }
+        Ok(())
+    }
+
+    fn checksum_commit_contents(
```

Comment:
IMO this looks ugly...why did rustfmt do that?

Reply:
I think because line is too long. Let me check RFC style

Reply:
https://github.com/rust-lang-nursery/rustfmt/blob/master/Configurations.md#function-arguments